include/xgboost/tree_model.h

```diff
@@ -320,6 +331,8 @@ class RegTree {
     fo->Write(dmlc::BeginPtr(stats_), sizeof(RTreeNodeStat) * nodes_.size());
   }
 
+  void Save(Json* out) const;
```

Comment:
What do you think about using a smart pointer (I think unique_ptr could work here but I am fine with shared_ptr too)?

I think usage of smart pointers could make our code a bit safer. What do you think?

Reply:
It's pretty safe for current purpose.  The Json object is a stack allocated object in C_API, going through the model and then expired due to being out of scope.

Reply:
I forgot about object on the stack. Yes. You are right, raw pointers are the best solution in this case

Reply:
Well references are an even better option in this case...

Reply:
@AnthonyTruchet It's by google coding standard, just pass a pointer if it's meant to be changed.  So ...